### ISSUE TITLE
Propagate groups from upstream OIDC provider

### ISSUE THREAD

**Issue by furuholm** [coreos/dex#1065]
The OIDC connectors currently discards the groups claim. I would like this to be propagated.

**Comment by ericchiang** [comment]
This is blocked on https://github.com/coreos/dex/issues/863

Since so many providers implement refresh tokens differently, we don't actually re-query the upstream provider when a dex client refresh its token. So we don't have any way of keeping the groups list fresh.

**Comment by furuholm** [comment]
I see. But how does groups differ from the other properties that are propagated in the current version (username, email, email verified)? Does those never change?

**Comment by ericchiang** [comment]
Groups are much more likely to change dynamically. I'm not saying that it's good that we don't update the other claims, but we definitely want to address this before expanding our features to groups

**Comment by furuholm** [comment]
Makes sense. I can use a fork until #863 is solved.

Closing this issue. Feel free to reopen if you want to track this as something to do after #863 has been solved.

**Comment by ericchiang** [comment]
Yeah, going to re-open for tracking this issue. Thanks for opening and sorry we don't have a better answer today.

**Comment by jacksontj** [comment]
@ericchiang I have recently run into this a few times, would you guys be open to merging an option to pull groups? I understand that they don't get update when using a refresh token, but as it stands nothing else does either. My proposal would be to (1) add an option to enable/disable groups for OIDC or (2) allow it to be enabled through config (as is) -- either way it sounds like we should specifically call out the behavior in the docs.

**Comment by jacksontj** [comment]
I've opened #1434 with my proposed workaround.

**Comment by Morriz** [comment]
Any updates here? Concourse forked dex and managed to pull in groups. I am with @jacksontj on this one: offer documentation and recommendations (if using groups set short ttl).

**Comment by danielsand** [comment]
Hey Guys,
we really would love to see this happen. <3

**Comment by dskatz** [comment]
Now that Dex v.2.21.0 has been released with both Refresh Tokens and Groups claims, is this still an issue?  And if not, could the flag "insecureEnableGroups" be renamed  "enableGroups"?   
(Referenced PRs: #1180, #1185, #1434)


In the use-cases that I am testing, groups claims are refreshed pretty quickly < 3 minutes in most cases which satisfies requirements.  From my perspective there does not seem to be a security issue anymore.

**Comment by dene14** [comment]
@dskatz any guidance how you enabled groups? for me Google always returns `groups=[]`
Thanks in advance!

**Comment by lbvffvbl** [comment]
Guys. I'm trying to enable groups for google. Was added serviceAccountFilePath, adminEmail, insecureEnableGroups: true, but I'm still getting groups=[]. Where am I wrong?

**Comment by lbvffvbl** [comment]
> Now that Dex v.2.21.0 has been released with both Refresh Tokens and Groups claims, is this still an issue? And if not, could the flag "insecureEnableGroups" be renamed "enableGroups"?
> (Referenced PRs: #1180, #1185, #1434)
> 
> In the use-cases that I am testing, groups claims are refreshed pretty quickly < 3 minutes in most cases which satisfies requirements. From my perspective there does not seem to be a security issue anymore.

Hi, can you show example of config file for enable google groups for Google?

**Comment by lbvffvbl** [comment]
> @dskatz any guidance how you enabled groups? for me Google always returns `groups=[]`
> Thanks in advance!

Dene14, I found. you need to ask extra-claim groups but you need to use a connector type google, not oidc
https://github.com/dexidp/dex/blob/master/Documentation/connectors/google.md

**Comment by abdulsalama** [comment]
hi folks, so does groups now works for oidc or not? I'm trying to use oidc with okta and I tried insecureEnableGroups: true but still getting groups=[]. any idea?

**Comment by JoelSpeed** [comment]
@abdulsalama Did you make sure to add the `groups` claim to the scope, as this will also be needed

**Comment by abdulsalama** [comment]
@JoelSpeed yes I did. I had this in the config:
          scopes:
            - openid
            - profile
            - email
            - groups

and I see that groups is being requested in the initial call.

**Comment by dskatz** [comment]
Hi @abdulsalama,

With Okta specifically, groups aren't returned in the idtoken unless you pay for that feature.  You will need to make a second call to the userinfo endpoint after requesting the claims initially.

**Comment by nabokihms** [comment]
@abdulsalama If you still have the problem, you can fix it specifying `getUserInfo: true` in [OIDC connector settings](https://github.com/dexidp/dex/blob/master/Documentation/connectors/oidc.md). 

With the enabled option, dex will make a second call to the useinfo endpoint like @dskatz suggest. I tested it yesterday.

**Comment by abdulsalama** [comment]
Thanks @dskatz and @nabokihms for the hints. I will give that a try and see if it works.

**Comment by mlopez-eb** [comment]
> @abdulsalama If you still have the problem, you can fix it specifying `getUserInfo: true` in [OIDC connector settings](https://github.com/dexidp/dex/blob/master/Documentation/connectors/oidc.md).
> 
> With the enabled option, dex will make a second call to the userinfo endpoint like @dskatz suggest. I tested it yesterday.

hi @nabokihms & @dskatz . If I'm installing Concourse through Helm, where exactly would I edit this `getUserInfo: true`? I know this is for the dex configuration, but I don't see those options in the values.yaml for the Helm installation.

It's become a nuisance not being able to get a user's groups through OIDC or OAUTH =[

**Comment by lentzi90** [comment]
https://github.com/dexidp/dex/pull/1634
Does this mean we can close this? The readme also says that the OIDC connector supports group claims now.

**Comment by lukasmrtvy** [comment]
@lentzi90 you still need to use `insecureEnableGroups`, even if you are using `claimMapping` , but group are supported, at least in native oidc connector.

**Comment by kszymans** [comment]
> Guys. I'm trying to enable groups for google. Was added serviceAccountFilePath, adminEmail, insecureEnableGroups: true, but I'm still getting groups=[]. Where am I wrong?

Hey.
I have the exact problem and I've checked:
- Google's IAM service account key is accessible for dex
- Client claims `groups` attribute from dex

Although I gets empty groups list. I also gets no groups claim in Dex's approval screen:
![image](https://user-images.githubusercontent.com/68637706/168739533-85082b9f-75c9-461d-a4a0-0b81100d46fe.png)

Maybe the issue is somewhere in claims

**Comment by almereyda** [comment]
Could the issue be rescoped to also include upstream OAuth2 login flows? We are interested in passing on `groups` provided by GitHub, which is an OAuth2 connector (only) and would not be covered by the OP.

**Comment by afirth** [comment]
@almereyda it seems it's in already: https://github.com/dexidp/dex/blob/master/connector/github/github.go#L317

haven't used it yet, hoping to get this working soon
